File: HomeStack.java
public class HomeStack extends CardStack {
	private Suits suit;

	public HomeStack() {
	}

	public void setCoords() {
		if (size() != 0) {
			get(size() - 1).setX(30 + (3 + arrayIndex) * 130);
			get(size() - 1).setY(30);
		}
	}

	public Suits getSuit() {
		return suit;
	}

	public boolean validPlacement(Card base, Card toPlace) {
		if (toPlace == null)
			return false;
		if (suit == null && toPlace.getRank() == Ranks.ace)
			return true;
		if (base == null)
			return false;
		return base.compareTo(toPlace) == 1 && toPlace.getSuit() == suit;
	}
	
	public void reset() {
		suit = null;
	}
	
	public Card push(Card c) {
		if (suit == null)
			suit = c.getSuit();
		return super.push(c);
	}
	
	public Card pop() {
		if (empty())
			return super.pop();
		if (peek().getRank() == Ranks.ace)
			reset();
		return super.pop();
	}
	
	public void populate() {
		for(int i = 0; i < 18; i++)
			for(int j = 0; j < 10; j++)
				Util.stackMap[3 + 13 * (arrayIndex + 3) + j][3 + i] = this;
	}

}

File: solitaire_js/src/HomeStack.js
class HomeStack extends CardStack {
  constructor() {
    this.reset();
  }

  reset() {
    this.suit = null;
  }

  setCoords() {
    if (this.length > 0) {
      this.peek().setX(30 + (3 + this.indexOfStack) * 130);
      this.peek().setY(30);
    }
  }

  getSuit() {
    return this.suit;
  }

  // override
  validPlacement(base, toPlace) {
    if (toPlace == null)
      return false;
    else if (this.suit == null && toPlace.getRank() == 'ace')
      return true;
    else if (base == null)
      return false;
    else
      return (base.compareTo(toPlace) == 1 && toPlace.getSuit() == this.suit);
  }

  // override
  push(c) {
    if (suit == null)
      suit = c.getSuit();
    return super.push();
  }

  // override
  pop() {
    if (this.length == 0 || this.peek().getRank() == 'ace')
      this.reset();
    return super.pop();
  }
  
  // override
  populate(stackMap) {
    for (let i = 0; i < 18; i++)
      for (let j = 0; j < 10; j++)
        stackMap[3 + 13 * (this.indexOfStack + 3) + j][3 + i] = this;
  }
}